xds/src/main/java/com/linecorp/armeria/xds/client/endpoint/RouteEntryMatcher.java

```diff
@@ -0,0 +1,357 @@
+                    }
+                    if (allHeaders.size() == 1) {
+                        // happy path for most cases
+                        return stringMatcher.match(allHeaders.get(0));
+                    }
+                    final String joined = COMMA_JOINER.join(allHeaders);
+                    return stringMatcher.match(joined);
+                };
+            } else {
+                throw new IllegalArgumentException("Unsupported header matchCase: " + matchCase + '.');
+            }
+        }
+
+        boolean matches(ClientRequestContext ctx) {
+            final HttpRequest req = ctx.request();
+            final HttpHeaders headers;
+            if (req == null) {
+                headers = HttpHeaders.of();
+            } else {
+                headers = req.headers();
+            }
+            return matcher.test(ctx, headers) != headerMatcher.getInvertMatch();
+        }
+    }
+
+    static class PathMatcherImpl {
+
+        private final Predicate<ClientRequestContext> predicate;
+
+        PathMatcherImpl(RouteMatch routeMatch) {
+            final PathSpecifierCase pathSpecifierCase = routeMatch.getPathSpecifierCase();
+            final boolean caseSensitive = routeMatch.getCaseSensitive().getValue();
+            if (pathSpecifierCase == PathSpecifierCase.PREFIX) {
+                final StringMatcher matcher = StringMatcher.newBuilder()
+                                                           .setPrefix(routeMatch.getPrefix())
+                                                           .setIgnoreCase(!caseSensitive)
+                                                           .build();
+                final StringMatcherImpl matcherImpl = new StringMatcherImpl(matcher);
+                predicate = ctx -> matcherImpl.match(ctx.path());
+            } else if (pathSpecifierCase == PathSpecifierCase.PATH) {
+                final StringMatcher matcher = StringMatcher.newBuilder()
+                                                           .setExact(routeMatch.getPath())
+                                                           .setIgnoreCase(!caseSensitive)
+                                                           .build();
+                final StringMatcherImpl matcherImpl = new StringMatcherImpl(matcher);
+                predicate = ctx -> matcherImpl.match(ctx.path());
+            } else if (pathSpecifierCase == PathSpecifierCase.SAFE_REGEX) {
+                final StringMatcher matcher = StringMatcher.newBuilder()
+                                                           .setSafeRegex(routeMatch.getSafeRegex())
+                                                           .build();
+                final StringMatcherImpl matcherImpl = new StringMatcherImpl(matcher);
+                predicate = ctx -> matcherImpl.match(ctx.path());
+            } else if (pathSpecifierCase == PathSpecifierCase.CONNECT_MATCHER) {
+                predicate = ctx -> ctx.method() == HttpMethod.CONNECT;
```

Comment:
> Note that this will not match other Extended CONNECT requests (WebSocket and the like) as they are normalized in Envoy as HTTP/1.1 style upgrades.

I understand that the `CONNECT_MATCHER` implementation does not follow the Envoy's specification, as `CONNECT` is used for WebSocket in Armeria.

Reply:
> CONNECT is used for WebSocket in Armeria.

I don't really see a difference since for HTTP1 `CONNECT` isn't used either (and http2 websockets aren't supported in envoy).
https://github.com/line/armeria/blob/6e479a3bc0cce836e8c58bedab8574800ec05a9c/core/src/main/java/com/linecorp/armeria/client/websocket/DefaultWebSocketClient.java#L146

Nuances on the request state may differ, but the condition is more or less the same.
- https://github.com/envoyproxy/envoy/blob/f5bc3c5fd2594e30f912ce2b3e85b297d7412b75/source/common/router/config_impl.cc#L1500
- https://github.com/envoyproxy/envoy/blob/f5bc3c5fd2594e30f912ce2b3e85b297d7412b75/source/common/http/header_utility.cc#L247